eng/common/scripts/Verify-AgentOS.ps1

```diff
@@ -0,0 +1,34 @@
+param (
+    [Parameter(Mandatory = $true)]
+    [ValidateNotNullOrEmpty()]
+    [string] $AgentPool,
+
+    [Parameter(Mandatory = $true)]
+    [ValidateNotNullOrEmpty()]
+    [string] $AgentImage
+)
+
+$lookup = @(
+    @{ AgentPool = "Azure Pipelines"; AgentImage = "windows-2019"; AgentOSMatch = $IsWindows }
+    @{ AgentPool = "Azure Pipelines"; AgentImage = "vs2017-win2016"; AgentOSMatch = $IsWindows }
+    @{ AgentPool = "Azure Pipelines"; AgentImage = "ubuntu-latest"; AgentOSMatch = $IsLinux }
+    @{ AgentPool = "Azure Pipelines"; AgentImage = "ubuntu-20.04"; AgentOSMatch = $IsLinux }
+    @{ AgentPool = "Azure Pipelines"; AgentImage = "ubuntu-18.04"; AgentOSMatch = $IsLinux }
+    @{ AgentPool = "Azure Pipelines"; AgentImage = "ubuntu-16.04"; AgentOSMatch = $IsLinux }
+    @{ AgentPool = "Azure Pipelines"; AgentImage = "macOS-latest"; AgentOSMatch = $IsMacOs }
+    @{ AgentPool = "Azure Pipelines"; AgentImage = "macOS-10.05"; AgentOSMatch = $IsMacOs }
+    @{ AgentPool = "Azure Pipelines"; AgentImage = "macOS-10.04"; AgentOSMatch = $IsMacOs }
+    @{ AgentPool = "azsdk-pool-mms-ubuntu-1604-general"; AgentImage = "MMSUbuntu16.04"; AgentOSMatch = $IsLinux }
+    @{ AgentPool = "azsdk-pool-mms-ubuntu-1804-general"; AgentImage = "MMSUbuntu18.04"; AgentOSMatch = $IsLinux }
+    @{ AgentPool = "azsdk-pool-mms-ubuntu-2004-general"; AgentImage = "MMSUbuntu20.04"; AgentOSMatch = $IsLinux }
+    @{ AgentPool = "azsdk-pool-mms-win-2019-general"; AgentImage = "MMS2019"; AgentOSMatch = $IsWindows }
+)
+
+foreach ($mapping in $lookup)
+{
+    if (($mapping.AgentPool -eq $AgentPool) -and ($mapping.AgentImage -eq $AgentImage) -and $mapping.AgentOSMatch) {
+        return;
+    }
+}
+
+throw "Agent operating system failed validation!"
```

Comment:
Might be helpful to print what the actual and expected values were on failure here.

Reply:
That required a bit more scripting than I expected. What I ended up with was a rule type (to make dumping the rules easier) and changing the way I evaluate whether the OS is what we expected. Because OS detection is murky I was trying to rely on the ```$IsLinux``` etc variables. In the end I just set the AgentOS on each rule and then have a conditional that works with the appropriate ```$Is[os]``` variables to make the determination. When a failure occurs I dump out the OS information provided by .NET runtime/interop which is actually better that the ```$PSVersionTable``` values provided.